### ISSUE TITLE
Though I refer  #3320 and  #3300  i can't add my publishing profile 

### ISSUE THREAD

**Issue by sujeongHeo** [microsoft/BotFramework-Composer#3338]
## Describe the bug
Though I refer  #3320 and  #3300  i can't add my publishing profile 

## Version
BotFramework-Composer-1.0.1-linux-x86_64.AppImage

## Browser

- [ ] Electron distribution
- [ O ] Chrome
- [ ] Safari
- [ ] Firefox
- [ ] Edge

## OS

- [ ] macOS
- [ ] Windows
- [O] Ubuntu

## To Reproduce

First way
1. Generate publish config with the provisionComposer.js
2. Create a new profile
3. Paste the configuration, choose Publish bot to an Azure Web App (preview) and give a name
4. Click Save
(Click OK on this error message #3320)
Click "Publish to selected profile"

Second way
1. i also tried to deleting some elements of profile that are not in the given format.

 for example,  the given format of 'Publish bot to an Azure Web App(preview) look like this. 

`  "accessToken": "<Access token from az account get-access-token>",
  "name": "<unique name in your subscription>",
  "environment": "dev",
  "settings": {
    "applicationInsights": {
      "InstrumentationKey": "<Instrumentation Key>"
    },`

But in my publish profile, there is a "hostname". so i didn't paste it .

2. it also not works

3. Third way 

1. I tried to make a file in /BotFrameworkComposer/publishBots/sth/deployment.appsettings.json.

2. and I just copy the one that i get from configuring provisionComposer.js, and paste on appsettings.json file that i made above

3. then I just click the button. 'Publish to selected profile'
![image](https://user-images.githubusercontent.com/41574104/83857765-6878e180-a756-11ea-9892-198d3cc58fb4.png)
But it doesn't work too

4. But it doesn't work too.

## Expected behavior

Publish success

## Screenshots

![image](https://user-images.githubusercontent.com/41574104/83857978-bd1c5c80-a756-11ea-8a22-5a1d66d7b16a.png)
![image](https://user-images.githubusercontent.com/41574104/83858023-cd343c00-a756-11ea-8798-ea4fe9059810.png)

**Comment by zxyanliu** [comment]
Hi @sujeongHeo, did you have a chance to read this [doc](https://docs.microsoft.com/en-us/composer/how-to-publish-bot) about how to publish a bot to Azure?

**Comment by arafattehsin** [comment]
Hi @zxyanliu - I have the similar problem. I went through the docs! 
![image](https://user-images.githubusercontent.com/16351038/83946744-0acbbe80-a856-11ea-98b4-7e066ea61575.png)

**Comment by arafattehsin** [comment]
Just to add more information. While I was adding my details; I also noticed some required values which is neither mentioned in the docs (https://docs.microsoft.com/en-us/composer/how-to-publish-bot) nor we are aware what to put in that (except for Subscription ID) 

![image](https://user-images.githubusercontent.com/16351038/83946816-76ae2700-a856-11ea-882e-2a9c875a5956.png)

**Comment by zxyanliu** [comment]
@sujeongHeo and @arafattehsin Thanks for reporting the issue. We will look into the issue and update the docs soon if there is anything missing since we last tested and wrote the publish process.

**Comment by arafattehsin** [comment]
@zxyanliu Can you please update as what should I put in these values?

`{
"subscriptionID": "subscription ID of Azure",
"publishName": "",
"provision": ""
"name":""
}`

**Comment by sujeongHeo** [comment]
> Hi @sujeongHeo, did you have a chance to read this [doc](https://docs.microsoft.com/en-us/composer/how-to-publish-bot) about how to publish a bot to Azure?

@zxyanliu  Yeah, I also went through docs,
 Ok, I'll wait for docs be updated :)

**Comment by zxyanliu** [comment]
FYI. I just tested publishing a bot to Azure Web App and succeeded. 

Here is a gif for your reference: 

![publish](https://user-images.githubusercontent.com/32497439/83975152-20b59f80-a8a7-11ea-9e27-2d295fd2b159.gif)

Thanks!

**Comment by zxyanliu** [comment]
@luhan2017 do you know what I should update in the doc for these? 

> @zxyanliu Can you please update as what should I put in these values?
`{
"subscriptionID": "subscription ID of Azure",
"publishName": "",
"provision": "",
"name":""
}`

**Comment by arafattehsin** [comment]
Just to update the team that my issue is not about primarily creating / provisioning the resources from the script. My issue is that I already have the below resources and I want to what should I put in which tag? I've already created. 

1. Bot Channel Registration
2. Web App 

So the question is, how do I map my bot using Composer to this? The manifest should have some good description. 

![image](https://user-images.githubusercontent.com/16351038/84027942-27d9cc00-a9d3-11ea-9e06-cf01698f89e6.png)

This is what I am getting.

**Comment by hibrenda** [comment]
@VanyLaw @zidaneymar Could you help to provide more info ?

**Comment by luhan2017** [comment]
@arafattehsin 

If you just want to deploy your bot to ABS without configuring any other services, you can just use a profile as the following:

{
  "accessToken": "<your access token>",
  "hostname": "<your web app name>",  
  "settings": {
    "MicrosoftAppId": "<the appid of your bot channel registration>",
    "MicrosoftAppPassword": "<the app password of your bot channel registration>"
  }
}

If you have luis configured in your composer bot, you should use this:
{
  "accessToken": "<your access token>",
  "hostname": "<your web app name>",
  "luisResource": "<your luis service name>", 
  "settings": {
    "luis": {
      "endpointKey": "<your luis endpointKey>",
      "authoringKey": "<your luis authoringKey>",
      "region": "<your luis region, for example westus>"
    },
    "MicrosoftAppId": "<the appid of your bot channel registration>",
    "MicrosoftAppPassword": "<the app password of your bot channel registration>"
  }
}

Or, you can always deploy your bot in visual studio by:
1. copy your bot assets to ComposerDialogs folder in the runtime/azurewebapp or runtime/azurefunctions folder. 
2. update the appsettings.deployment.json 

Please let me know if it works for you or not. 

@zxyanliu let's work together offline to provide more info for provision profile in the doc .

**Comment by luhan2017** [comment]
@sujeongHeo For your screenshot 1, this is a known issue, and it won't effect anything, just click OK to work around it. For your screenshot 2, I am wondering if you are using vscode so that composer have issues updating that file, could you close that and try again?

Please let me know if it still not works.

**Comment by arafattehsin** [comment]
@luhan2017 Yes I do have a LUIS associated with my Composer bot and I am really sorry but that was not helpful for me. I can understand why others are also really puzzled with this. Now let me write down something super simple for you and then you may advise what needs to be done. 

![image](https://user-images.githubusercontent.com/16351038/84148101-d2b8bb80-aaa1-11ea-8fff-d188587c6ecf.png)

My LUIS Application name in **LUIS Portal:** AMABank(composer)-amabank.en-us.lu
My LUIS Resource name in **Azure Portal**: luisdemoservice-Authoring
My App Service (Web App) name created in **Azure Portal**: amabank
My Bot Channel Registration in **Azure Portal**: AMABank

**Question # 1:** When you write **endpointKey**, where do you get it from? 
**Question # 2:** When you write **authoringKey**, where do you get it from?

I could only see the endpoint key in the LUIS Portals;

![Annotation 2020-06-09 223934](https://user-images.githubusercontent.com/16351038/84149816-63909680-aaa4-11ea-81b2-ee44d00f46d9.png)

![Annotation 2020-06-09 224249](https://user-images.githubusercontent.com/16351038/84149843-6f7c5880-aaa4-11ea-8538-a0aa1a13e695.png)

From my Azure Portal, here's the snapshot: 

![image](https://user-images.githubusercontent.com/16351038/84150150-ceda6880-aaa4-11ea-8c84-63ad9305576e.png)

**Coming to your Visual Studio publish part:**

Unfortunately, I was not able to locate ComposerDialogs in my runtime folder which is there in %appdata\botframeworkcomposer\hostedBots% - I think another detailed walkthrough is needed there later on (in the docs hopefully). 

**My Request:** I want to publish the Bot through Composer (also as a skill) so that I can demonstrate this to the audience that it really works. Appreciate your help in that!

**Comment by luhan2017** [comment]
@arafattehsin, 

First, for the get address info error in your deployment, I suspect it is because the key and endpoint not match. It is a little bit tricky here, let me explain. 

Previously in composer, we auto generated the luis authoring and prediction endpoint by this pattern: `https://${luis.region}.api.cognitive.microsoft.com` which works fine for a while.

But recently, there is a change in the luis side which update the naming pattern as this:
`https://{your-resource-name}.api.cognitive.microsoft.com`
see https://docs.microsoft.com/en-us/azure/cognitive-services/luis/developer-reference-resource as a reference. 

We updated composer side and expose another field called "luis.endpoint" to fix the above issue, the fix is, if luis:endpoint is not null, just use luis.endpoint as the authoring and prediction endpoint.

But our fix missed one possible scenario like you reported, only authoring endpoint is migrated to this pattern, https://{your-resource-name}.api.cognitive.microsoft.com, the prediction endpoint stays the same as before https://${luis.region}.api.cognitive.microsoft.com.

We are working on another fix #3364 as a P0 item, hope this will unblock you. After this fix is in, you can configure your settings like this, sorry for the inconvenience, will keep you updated. 

```
{
  "accessToken": "<your access token>",
  "hostname": "<your web app name>",
  "luisResource": "<your luis service name>", 
  "settings": {
    "luis": {
      "endpointKey": "<your luis endpointKey>",
      "endpoint": "<your luis prediction endpoint>" // if this value is null, we will generate this https://${luis.region}.api.cognitive.microsoft.com
      "authoringKey": "<your luis authoringKey>",
      "authoringEndpoint": "<your luis authoring endpoint>" // if this value is null, we will generate this https://${luis.region}.api.cognitive.microsoft.com
      "region": "<your luis region, for example westus>"
    },
    "MicrosoftAppId": "<the appid of your bot channel registration>",
    "MicrosoftAppPassword": "<the app password of your bot channel registration>"
  }
}
```

**Comment by arafattehsin** [comment]
@luhan2017 Thank you very much for this detailed explanation on what's going on behind the scenes. May I know which key do you refer as endpointKey and luisResource? It'd be great if I can get the screenshot or you can refer mine. 

I just retried with your configuration and it gives me this error;

![image](https://user-images.githubusercontent.com/16351038/84258740-c6446980-ab5a-11ea-9fe3-9cdb33ec9bbd.png)

```
{
  "accessToken": "access token from azure",
  "name": "amabank",
  "settings": {
    "luis": {
      "endpointKey": "_authoring key from the screenshot 1_",
      "authoringKey": "_authoring key from the screenshot 2_",
      "region": "westus",
      "endpoint": "https://westus.api.cognitive.microsoft.com/",
      "authoringEndpoint": "https://luisdemoservice-authoring.cognitiveservices.azure.com/"
    },
    "MicrosoftAppId": "my app id",
    "MicrosoftAppPassword": "my app password"
  },
  "hostname": "amabank",
  "luisResource": "luisdemoservice-authoring"
}
```

**Screenshot 1**

![image](https://user-images.githubusercontent.com/16351038/84259015-30f5a500-ab5b-11ea-8bac-cf040d908e23.png)

**Screenshot 2**

![image](https://user-images.githubusercontent.com/16351038/84259214-7b772180-ab5b-11ea-826a-062b3813a172.png)

**Comment by arafattehsin** [comment]
@luhan2017 No worries. I appreciate what you all as a team are doing. As I could see your PR - I'd like to make one more comment on it. You talked about the endpoint and authoring urls separately but are you considering the regions as well? 

My prediction resource and authoring resource both have got different regions.

**Comment by luhan2017** [comment]
@arafattehsin , thanks for your feedback, but there is a limitation in luis that you must author and publish in the same region, that is why we keep the one region setting. please refer to this doc https://docs.microsoft.com/en-us/azure/cognitive-services/luis/luis-reference-regions for more details. 

I've just merge the PR in master, could you please check if your scenario works in master now? Thank you very much for assistance and patience on resolving this issue.

**Comment by arafattehsin** [comment]
@luhan2017 Sure. Totally understood the region part. I will check and give it a go! 🤘

**Comment by arafattehsin** [comment]
@luhan2017 This is to update you that it works absolutely fine now. I am also writing a blog post on this so folks who come across this issue can easily follow the guidelines. Appreciate your help. 💯

**Comment by zxyanliu** [comment]
FYI. The doc has been updated in [PR 133](https://github.com/MicrosoftDocs/composer-docs-pr/pull/133) which just got approved. The content will go live soon. Thanks again for your feedback.

**Comment by arafattehsin** [comment]
Thanks @zxyanliu and @luhan2017 - I also wrote some of the detailed stuff here for our creators! - https://www.arafattehsin.com/bot-framework-composer-publish-azure-resources/

**Comment by hibrenda** [comment]
addressed in #3364
